python-checks/src/main/java/org/sonar/python/checks/NumpyListOverGeneratorCheck.java

```diff
@@ -0,0 +1,100 @@
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.HasSymbol;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.cfg.fixpoint.ReachingDefinitionsAnalysis;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6714")
+public class NumpyListOverGeneratorCheck extends PythonSubscriptionCheck {
+
+  public static final String MESSAGE = "Pass a list to \"np.array\" instead of passing a generator.";
+  private ReachingDefinitionsAnalysis reachingDefinitionsAnalysis;
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> reachingDefinitionsAnalysis = new ReachingDefinitionsAnalysis((ctx.pythonFile())));
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, this::checkNumpyArrayCall);
+  }
+
+  private void checkNumpyArrayCall(SubscriptionContext ctx) {
+    CallExpression call = (CallExpression) ctx.syntaxNode();
+    Optional.ofNullable(call.calleeSymbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter("numpy.array"::equals)
+      .ifPresent(fqn -> checkGeneratorCallee(call, ctx));
+  }
+
+  private void checkGeneratorCallee(CallExpression call, SubscriptionContext ctx) {
+    List<Argument> argList = call.arguments();
+    if (argList.isEmpty()) {
+      return;
+    }
+
+    if (Optional.of(argList.get(0))
+      .filter(arg -> arg.is(Tree.Kind.REGULAR_ARGUMENT))
+      .map(RegularArgument.class::cast)
+      .map(RegularArgument::expression)
+      .filter(regArg -> (regArg.is(Tree.Kind.GENERATOR_EXPR) || this.isNamedGeneratorExpression(regArg)))
+      .isEmpty()) {
+      return;
+    }
+
+    if (Optional.ofNullable(TreeUtils.nthArgumentOrKeyword(1, "dtype", argList))
+      .filter(regArg -> regArg.expression().is(Tree.Kind.NAME))
+      .map(regArg -> (Name) regArg.expression())
+      .map(HasSymbol::symbol)
+      .map(Symbol::fullyQualifiedName)
+      .filter("object"::equals)
+      .isEmpty()) {
+      ctx.addIssue(call, MESSAGE);
+    }
+  }
+
+  private boolean isNamedGeneratorExpression(Expression expression) {
+    return Optional.of(expression)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(Name.class))
+      .map(name -> this.reachingDefinitionsAnalysis.valuesAtLocation(name))
+      .filter(NumpyListOverGeneratorCheck::checkSetProperties)
+      .isPresent();
+  }
+
+  private static boolean checkSetProperties(Set<Expression> set) {
+    return !set.isEmpty() && set.stream().allMatch(NumpyListOverGeneratorCheck::getSetCondition);
+  }
+
+  private static boolean getSetCondition(Expression expression) {
```

Comment:
Here, I think the name could be a bit more explicit i.e isGeneratorAndParentHasType or something like that.

Reply:
Changed in 3fb81d1c3966f5db1a9807c217e1f7a94173e340.